core/src/main/java/io/questdb/griffin/engine/groupby/SampleByFirstLastRecordCursorFactory.java

```diff
@@ -0,0 +1,755 @@
+                        newState = STATE_OUT_BUFFER_FULL;
+                    } else if (samplePeriodIndexOffset - prevSamplePeriodOffset == maxSamplePeriodSize - 1) {
+                        //  search came to the end of sample by by periods
+                        // re-fill periods and search again
+                        newState = STATE_SEARCH;
+                    } else {
+                        // Data frame exhausted. Next time start from fetching new data frame
+                        newState = STATE_FETCH_NEXT_DATA_FRAME;
+                    }
+
+                    if (rowsFound > 1) {
+                        record.of(currentRow = 0);
+                        newState = -newState;
+                    }
+
+                    // No rows to iterate, return where to continue from (fetching next data or index frame)
+                    return newState;
+
+                case STATE_RETURN_LAST_ROW:
+                    if (crossRowState != NONE) {
+                        record.of(currentRow = 0);
+                        // Signal there is a row by returning negative value
+                        return -STATE_DONE;
+                    }
+                    // Fall through to STATE_DONE;
+
+                case STATE_DONE:
+                    return STATE_DONE;
+
+                default:
+                    throw new UnsupportedOperationException("Invalid state " + state);
+            }
+        }
+
+        private boolean hasNext0() {
+            // Check if values from the last found sample by row have to be saved in crossFrameRow
+            checkCrossRowAfterFoundBufferIterated();
+            rowsFound = 0;
+
+            while (state != STATE_DONE) {
+                state = getNextState(state);
+                if (state < 0) {
+                    state = -state;
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private void saveFirstLastValuesToCrossFrameRowBuffer() {
+            // Copies column values of all columns to cross frame row buffer for found row with index 0
+            long firstRowId = firstRowIdOutAddress.get(0);
+            long lastRowId = lastRowIdOutAddress.get(0);
+            for (int i = 0; i < isColumnLast.length; i++) {
+                if (i == groupByTimestampIndex) {
+                    long tsIndex = startTimestampOutAddress.get(0) - prevSamplePeriodOffset;
+                    crossFrameRow.set(i, samplePeriodAddress.get(tsIndex));
+                } else {
+                    long rowId = isColumnLast[i] ? lastRowId : firstRowId;
+                    saveRowIdValueToCrossRow(rowId, i);
+                }
+            }
+        }
+
+        private void saveFixedColToBufferWithLongAlignment(int index, DirectLongList saveBufferAddress, int columnType, long pageAddress, long rowId) {
+            switch (ColumnType.pow2SizeOf(columnType)) {
+                case 3:
```

Comment:
magic numbers

Reply:
Not so magic, our column sizes in power of 2. Alternative will be to switch every column type which is more code for nothing